### PR TITLE
Add circular genome support

### PR DESCRIPTION
**Pull request by SJShaw**
This somewhat massive PR adds circular genome support to almost all detection and analyses. Some parts are not quite complete but the changes are at a stage that other people should experiment with it and see what they think of the current setup.

**NOTE:** As of opening the PR, it includes an embedded version of the javascript for the HTML output for experimental purposes. This will always take precedence over those fetched with `download-antismash-databases`, and should not be merged as is. 

General performance optimisations haven't been looked for as yet. Some style issues have also been left in, as the final clean up can wait until the implementation in general is acceptable/accepted. More unit tests would not go astray and some reasonably complex integration test should also be added.

## Visualisation
Record overview drawings show wrapping around the ends of the genome, rather than using a great deal of space to show a circular view, e.g. this one that also shows a bug where a non-crossing region is duplicated:
image

Another example, with a small, arbitrarily created record:
image

Individual region drawings have split the axis coordinates at the origin to show the real coordinates, along with a vertical marker to make the origin more obvious, e.g. this shifted kirromycin:
image

Cross-origin genes/CDS features are shown as usual, but have the marker laid over the top (this is a rotated nisin):
image

For both regions and CDS features that cross the origin, coordinates are shown not from "lowest to highest" but "highest to lowest", e.g.:
image
image

## Other major changes
- **secmet:** records and features support crossing the origin, from individual CDS features to full regions
- **secmet:** `Record` objects have helper methods for extending locations in ways that are circular-aware 
- **secmet:** sorting of feature sets that collectively cross the origin sort pre-origin to post-origin
- **secmet:** formation of both candidate clusters and regions support cross-origin protoclusters
- **all_orfs:** gene detection can find genes spanning the origin
- **nrps_pks_domains:** supports both cross-origin domains and cross-CDS modules that also cross the origin
- **RiPP modules:** search across the origin for both precursors and other CDS features that contain domains of interest

 ## Known issues/unimplemented features
 - [x] **hmmdetection**: in some records, some protoclusters are duplicated
 - [x] **genefinding**: neither prodigal nor glimmerhmm (via `--genefinding-tool`) have been updated to find genes over the origin
 -  [X] **clusterblast:** visualisation has not been updated for cross-origin regions, **but** this has a separate update in #701 that needs to be merged first
 - [x] **clusterblast:** database build scripts have not been updated to handle cross-origin areas
 - [x] **clusterblast:** reference areas that cross the origin have no visualisation support
 - [x] **TFBS:** visualisation of cross-origin sites is not yet supported
 - [x] **TFBS:** visualisation of sites that don't cross the origin but where the nearest gene *is* on the other side of the origin is not yet handled
 - [x] **sideloader:** sideloaded subregions are not consistently drawn correctly in the visualisation
 - [x] **secmet:** per-region genbank outputs for cross-origin regions aren't quite complete, this will need some movement of coordinates to not cross the origin or some other input-modifying change
- [x] **HTML:** in the record overview SVG, regions on a contig-edge in a linear genome will also create a single pixel "region" on the other end of the genome
- [x] **HTML:** browser performance is poor on the overview page when multiple circular genomes have cross-origin regions
- [ ] **HTML:** region SVG axis ticks can sometimes be awkwardly placed, though this may not be able to be resolved for all cases
- [x] **HTML:** NCBI genomic context links can't handle cross-origin genes, they may need to be removed for those genes in particular
- [x] **CASSIS:** has not been updated at all, but circular fungal genomes aren't expected

There are likely more, but more inputs need to be tested to find those bugs/edge cases that I've missed.

### CODE REVIEW

**Comment by SJShaw** [comment]
> Nice feature; I like the HTML visualizations.
> 
> I've tested it on a few Vibrio plasmids with the anguibactin BGC and default settings.
> 
> **AY312585.1** looked good.
>
> **CP020532.1** failed. I'm guessing because the BGC region spans the entire plasmid (fun new edge case!)

The cause of this one was a bug in reducing multiple locations into a single location due to an off-by-one error.
 
 I did then test it with some variants of full-contig clusters, and some did fail due to a different off-by-one error.
 
> **NZ_CP016262** has bad coordinates for the origin-spanning gene but otherwise looked good:

This affected only cross-origin genes in the reverse strand, it was a good catch.

They've all been fixed now, with regression tests added. Since I was rebasing these fixes in anyway, I also rebased on top of current master, apologies for breaking any existing checkouts.

**Comment by SJShaw** [comment]
Another issue I didn't mention but will need to be handled in some way is NCBI context links for cross-origin features. The links we currently use `...&id=<accession>&from=<start>&to=<end>` aren't useful for cross-origin features. When the start is greater than the end, the NCBI page will take the average of the two and show a single point, e.g. this 70kb conting with a start of 25kb and an end of 5kb:
image

**Comment by zreitz** [comment]
NCBI just doesn't seem to handle circular genomes very well.
image

Maybe using two iframes?
image

**Comment by kblin** [comment]
I don't think I want to host another site's content in an iframe on one of our pages. Especially as the NCBI sets some cookies that we'd probably have to go inform users about. I feel like if the NCBI can't show a proper genomic context for cross-origin CDSes, I'd just not have a link for those.

**Comment by SJShaw** [comment]
Rebased once more:
- NCBI genomic context link removed for cross-origin genes
- Record overview SVGs with a region on the contig edge no longer create a matching single pixel wide region on the other edge of the contig
- Some style fixes

The javascript addition is now at the end of the commit list, so future changes to the javascript will hopefully be less disruptive in terms of commit log spam

**Comment by SJShaw** [comment]
TFBS now supports cross-origin sites and finding neighbours over the origin.

Helpers have been added to CDS collections for the usual `.cds_children` access. Along with supporting all their existing use cases as a tuple, they now support fetching the subdivisions within that collection, e.g. `.cds_children.pre_origin`, `.cds_children.cross_origin`.

**Comment by SJShaw** [comment]
Both CASSIS and fungal genefinding via glimmerhmm have been explicitly disabled for circular inputs. While there's some circular plasmids in fungi, there's neither very many nor are fungal inputs assembled well enough for it to matter a great deal.

Prodigal genefinding over the origin has been added by extending the sequence with a repeat of the first 50kb of the input or 90% of the input, whichever is smaller.

Almost all uses of genefinding have FASTA files as inputs, so these are unlikely to be requested very often.

**Comment by SJShaw** [comment]
The javascript performance mentioned in the initial comment doesn't seem to be related to the changes in this PR. The entirety of the javascript execution is between 15ms and 80ms over a few runs of a dummy input of 6 circularised clusters, with both the javascript used by master and this PR

The full page load time is roughly 1 to 1.5 seconds, by comparison, for both version of the JS.

**Comment by SJShaw** [comment]
(Painfully) Rebased once more:
- commit history made much cleaner and broken into more self-contained chunks (where possible)
- more unit tests
- a global integration test for a rotated nisin cluster
- proper handling of region genbank outputs for cross-origin regions by first converting them to be linear

Some commits can also be split out into a smaller PR that isn't specific to circular inputs (e.g. location subclassing and splitting up the region code into sections).

**Comment by SJShaw** [comment]
Updated with clusterblast visualisation support for cross-origin query regions.
Also cleans up most of the style issues that previous had added.

**Comment by SJShaw** [comment]
Rebased on master once more.

Fixes a protocluster size issue that occurred only when multiple genes crossed the origin and each gene, taken individually, satisfied the conditions for same rule (e.g. most NRPS genes).

Adds an explicit rejection when attempting to use `--start` and `--end`  to define a cross-origin area. This is due to slicing over the origin will converting all locations on all results, rather than just those specific output files for which it is required (i.e region genbank outputs for regions that cross the origin).

**Comment by SJShaw** [comment]
Latest update fixes an issue where ClusterCompare drawing failed on `--reuse` due to drawing coordinates from database hits not being kept in result JSON.

**Comment by SJShaw** [comment]
Rebased on top of master once more, with the following updates:
- Clusterblast cross-origin visualisation support
- Clusterblast database generation supports cross-origin reference areas (and script moved from https://github.com/antismash/generate-databases to the module's `data` dir, similar to other modules)
- Removes Clusterblast's internal homology blast, as that hasn't been in use since before antiSMASH 5

After running on a batch of RefSeq circular genomes, I also found a couple of very specific assertion failures that I've fixed up and added tests for.

With these Clusterblast changes, that should be all affected data structures and modules caught up to supporting circular genomes. There's some included commits that are effectively unrelated and can move to a separate pull request. After that, this PR just needs some testing by some other people to make sure outputs make sense. It'd also be nice to test it on a whole lot more genomes on top of what I've tested with.

**Comment by SJShaw** [comment]
I ran the branch over 150 genomes from the antiSMASH database that had two regions on a contig edge _and_ had at least one circular contig. This found a few bugs and most have been fixed:
- TFBS crashed when finding intergenic areas for cross-origin regions when only post-origin or pre-origin genes were present
- candidate cluster formation failed sometimes when almost all of the contig was covered by protoclusters, including at least one over the origin (e.g. NZ_AP017310, a 100kb circular plasmid with 80kb covered by protoclusters)

There's three more issues that haven't yet been fixed:
- HTML: protocluster bars in region views occasionally overlap (e.g. NC_019746)
- a circular record entirely covered by a single region and at least one cross-origin area will fail due to existing checks, removing those checks needs to be tested against all the other parts that are changed to ensure it doesn't cause other, quieter problems

Another issue that's not just a bug, but completely unhandled is in HTML when a circular record entirely covered by more than one subregion/protocluster. Right now, (with the checks loosened as mentioned above and a manually annotated subregion to completely fill the plasmid) it looks like the following:
image

The NRPS and T1PKS clusters are cross-origin and an operon, but are broken into two and failing to draw. It would probably be best to rotate the visualisation to have the edges be in a neighbouring area. When not possible to split without interrupting a cluster core, some kind of indicator should be added so that it's easier for users to see that it's an interruption in visualisation only, not that it's two separate clusters.

**Comment by SJShaw** [comment]
- Rebased on top of master once more
- Revamps all cluster/area bar packing for HTML to fix issues above
- Corrected most style issues

Reworked all region visualisations to account for whole-contig regions with circular components. The previous visualisations for cross-origin regions remains the same, but the whole-contig versions now look like this example:

All cross-origin clusters, subregions, and genes within a non-cross-origin region now have circular markers. Selecting either side will also select the other side, so they behave exactly as non-split versions.
To make the gene table consistent, split genes have an entry both at the start of the table and at the end of the table. The intent here is that if a user is looking at the end of the table, it won't be misleading to have the last gene in the table the second last (or earlier, in the case of multiple cross-origin genes) gene in the visualisation, and vice versa for the start of the table.

A slightly contrived example, with a subregion added to connect the pre-origin and post-origin sides:
image


While the javascript is still embedded in the PR (and thus blocks it), I think this is the last update, and if review(s) don't find any other issues, then the hardcode can be replaced with a usual fetch via `download-antismash-databases`.

**Comment by SJShaw** [comment]
Slightly tweaked the HTML update tests to remove some unnecessary mocks.

**Review by kblin — COMMENTED** [review]
Whew, what a patch set. I only have a few comments inline.

File: antismash/common/secmet/test/test_secmet.py [hunk 1]
```diff
@@ -1008,11 +1044,12 @@ def test_simple_underflow(self):
 
         new = self.record.extend_location(location, distance)
         assert isinstance(new, CompoundLocation)
+        print(f"{location=} + {distance=} = {new=}")
```
[the remark this answers is not in the document]
Comment (SJShaw) [reply antismash/common/secmet/test/test_secmet.py]: Removed.

File: antismash/common/all_orfs.py [hunk 2]
```diff
@@ -145,9 +156,12 @@ def create_feature_from_location(record: Record, location: FeatureLocation,
     """
     if label is None:
         digits = len(str(len(record)))
-        label = 'allorf_{start:0{digits}}_{end:0{digits}}'.format(
-            digits=digits, start=(location.start + 1), end=location.end
-        )
+        if len(location.parts) > 1:  # crosses the origin
+            label = f"allorf_{location.parts[0].start}_{location.parts[1].end + 1}"
```
[the remark this answers is not in the document]
Comment (SJShaw) [reply antismash/common/all_orfs.py]: Definitely, it's been updated to add the padding.

File: antismash/common/test/test_all_orfs.py [hunk 3]
```diff
@@ -123,6 +129,29 @@ def test_multiframe(self):
...
+        area_seq = area.location.extract(record.seq)
+        assert area_seq == "NNNNNNNNNNNNNNNNNNNNNNNNNATGNNNNNNNNNNNNNNNNNNNNNTAGNNNNNNNN"
+        found_start = area_seq.find("ATG")
+        found_end = area_seq.find("TAG")
+        assert 0 <= found_start < found_end <= len(record)
+        assert (found_end - found_start) % 3 == 0, "not in the same frame"
+        # then check the full finding method
+        features = find_all_orfs(record, area, min_length=20)
+        assert len(features) == 1
+        expected_loc = CompoundLocation([FeatureLocation(52, 57, 1), FeatureLocation(0, 22, 1)])
```
[the remark this answers is not in the document]
Comment (SJShaw) [reply antismash/common/test/test_all_orfs.py:151]: Fixed, I probably confused myself with the lower number always being the start prior to the PR.

File: antismash/common/secmet/features/protocluster.py [hunk 4]
```diff
@@ -92,7 +92,7 @@ def definition_cdses(self) -> Set[CDSFeature]:
         """ Returns the set of CDSFeatures responsible for the creation of this protocluster """
         return set(self._definition_cdses)
 
-    def add_cds(self, cds: CDSFeature) -> None:
+    def add_cds(self, cds: CDSFeature, section: CollectionSection = None) -> None:
         super().add_cds(cds)
```
[the remark this answers is not in the document]
Comment (SJShaw) [reply antismash/common/secmet/features/protocluster.py]: It should, fixed.

File: antismash/common/secmet/features/region/structures.py [hunk 5]
```diff
@@ -114,7 +115,7 @@ def detection_rules(self) -> List[str]:
                 rules[product] = rule
         return list(rules.values())
 
-    def add_cds(self, cds: CDSFeature) -> None:
+    def add_cds(self, cds: CDSFeature, _section: CollectionSection = None) -> None:
         """ Adds a CDS to the Region and all relevant child collections. Links
             the CDS back to this region
         """
```
[the remark this answers is not in the document]
Comment (SJShaw) [reply antismash/common/secmet/features/region/structures.py:121]: Fixed as well as the protocluster case of the same thing.

**Comment by SJShaw** [comment]
Rebased once again on current master, including fixes at the same time.

Given there's no pending issues with HTML visualisation, the javascript embed has been removed and now uses the normal download system as the library is now versioned and downloadable.

**Review by kblin — APPROVED** [review]
Looks good to me 👍
